File: frontend/src/types/contest.ts
/** TypeScript types for Contest management domain. */

export interface ContestResponse {
  id: number;
  group_id: number;
  cf_contest_id: number;
  name: string;
  min_solved: number;
  min_solved_is_percent: boolean;
  total_problems: number;
  required_solved: number;
  participants: string[];
  lock_participants: boolean;
  start_date: string | null;
  last_refreshed: string | null;
  created_at: string | null;
  result_count: number;
}

export interface CachedResultResponse {
  id: number;
  contest_id: number;
  handle: string;
  solved_count: number;
  passed: boolean;
  participated: boolean;
  rank: number;
  cached_at: string | null;
}

export interface ContestResultsResponse {
  contest: ContestResponse;
  total: number;
  passed: number;
  failed: number;
  not_participated: number;
  results: CachedResultResponse[];
}

export interface ContestStandingsResponse {
  contest: ContestResponse;
  standings: CachedResultResponse[];
  summary: {
    total: number;
    participated: number;
    not_participated: number;
  };
}

export interface FetchHistoryResponse {
  id: number;
  contest_id: number;
  fetch_date: string;
  passed_count: number;
  failed_count: number;
  not_participated_count: number;
  total_count: number;
}

export interface TaskDispatchResponse {
  status: string;
  task_id: string;
  message: string;
}

export interface TaskStatusResponse {
  task_id: string;
  status: "PENDING" | "PROGRESS" | "SUCCESS" | "FAILURE";
  progress?: number;
  message?: string;
  result?: Record<string, unknown>;
  error?: string;
}

export interface ContestCreatePayload {
  group_id: number;
  cf_contest_id: number;
  name?: string;
  min_solved?: number;
  min_solved_is_percent?: boolean;
  participants?: string[];
  lock_participants?: boolean;
}

export interface BatchContestCreatePayload {
  group_id: number;
  contest_ids: number[];
  min_solved?: number;
  min_solved_is_percent?: boolean;
}

export type ResultStatusFilter =
  | "all"
  | "passed"
  | "failed"
  | "participated"
  | "not_participated";

File: backend/app/schemas/contest.py
"""Pydantic schemas for Contests, Results, Standings, and Progress."""
from typing import Any, Dict, List, Optional
from pydantic import BaseModel, Field


class ContestCreate(BaseModel):
    group_id: int
    cf_contest_id: int
    name: Optional[str] = None
    min_solved: int = Field(default=1, ge=0)
    min_solved_is_percent: bool = False
    participants: Optional[List[str]] = Field(default_factory=list)
    lock_participants: bool = False


class BatchContestCreate(BaseModel):
    group_id: int
    contest_ids: List[int] = Field(..., min_length=1)
    min_solved: int = Field(default=1, ge=0)
    min_solved_is_percent: bool = False


class ContestUpdate(BaseModel):
    name: Optional[str] = None
    min_solved: Optional[int] = None
    min_solved_is_percent: Optional[bool] = None
    participants: Optional[List[str]] = None
    lock_participants: Optional[bool] = None


class ContestResponse(BaseModel):
    id: int
    group_id: int
    cf_contest_id: int
    name: str
    min_solved: int
    min_solved_is_percent: bool
    total_problems: int
    required_solved: int
    participants: List[str]
    lock_participants: bool
    start_date: Optional[str] = None
    last_refreshed: Optional[str] = None
    created_at: Optional[str] = None
    result_count: int = 0

    class Config:
        from_attributes = True


class CachedResultResponse(BaseModel):
    id: int
    contest_id: int
    handle: str
    solved_count: int
    passed: bool
    participated: bool
    rank: int
    cached_at: Optional[str] = None

    class Config:
        from_attributes = True


class ContestResultsResponse(BaseModel):
    contest: Dict[str, Any]
    total: int
    passed: int
    failed: int
    not_participated: int
    results: List[CachedResultResponse]


class ContestStandingsResponse(BaseModel):
    contest: Dict[str, Any]
    standings: List[CachedResultResponse]
    summary: Dict[str, int]


class FetchHistoryResponse(BaseModel):
    id: int
    contest_id: int
    fetch_date: str
    passed_count: int
    failed_count: int
    not_participated_count: int
    total_count: int

    class Config:
        from_attributes = True


class TaskDispatchResponse(BaseModel):
    status: str = "QUEUED"
    task_id: str
    message: str
